File: wagerproof-ios-native/Wagerproof/Features/GameCards/Components/GameCardFormatting.swift
import Foundation
import SwiftUI
import WagerproofDesign

/// Shared formatting helpers + team color fallback used by the NFL and CFB
/// game cards / bottom sheets. Mirrors RN `utils/formatting.ts` +
/// `utils/teamColors.ts`.
///
/// FIDELITY-WAIVER #008: Real per-team color tables (the 500-line lookup in
/// `teamColors.ts`) port with the sport-specific batches. B04 uses a single
/// neutral palette so cards render even without team colors; reviewers
/// receive screenshots with placeholders highlighted.
enum GameCardFormatting {
    /// Formats a moneyline (+/-) integer to its display string.
    static func formatMoneyline(_ value: Int?) -> String {
        guard let value else { return "—" }
        return value > 0 ? "+\(value)" : "\(value)"
    }

    /// Formats a spread to display string. Spread is signed Double.
    static func formatSpread(_ value: Double?) -> String {
        guard let value else { return "—" }
        if value == value.rounded() {
            let v = Int(value)
            return v > 0 ? "+\(v)" : "\(v)"
        }
        return value > 0 ? "+\(String(format: "%.1f", value))" : String(format: "%.1f", value)
    }

    /// Round to nearest half (38.27 → 38.5). Used for O/U display.
    static func roundToNearestHalf(_ value: Double?) -> String {
        guard let value else { return "—" }
        let rounded = (value * 2).rounded() / 2
        if rounded == rounded.rounded() {
            return "\(Int(rounded))"
        }
        return String(format: "%.1f", rounded)
    }

    // Every formatter below is a static: the NFL/CFB detail heroes call
    // `formatCompactDate` + `convertTimeToEST` from `topRow`, which SwiftUI
    // re-evaluates on every scroll frame. Built inline this was up to 7-8
    // DateFormatter/ISO8601DateFormatter constructions per frame at 120Hz.
    // One formatter per format string (rather than one whose `dateFormat` is
    // reassigned in a loop) — a mutated shared formatter is not thread-safe.
    private static let etTimeZone = TimeZone(identifier: "America/New_York")

    private static let isoFractional: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static let isoPlain: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    private static func fixedFormatter(_ format: String, timeZone: TimeZone?) -> DateFormatter {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = timeZone
        f.dateFormat = format
        return f
    }

    /// Date-only parsers, tried in order (ET — these strings carry no zone).
    private static let dateParsers: [DateFormatter] = [
        fixedFormatter("yyyy-MM-dd", timeZone: etTimeZone),
        fixedFormatter("yyyy-MM-dd HH:mm:ss", timeZone: etTimeZone)
    ]

    /// Time-only parsers, tried in order (UTC — these are wall-clock UTC).
    private static let timeParsers: [DateFormatter] = [
        fixedFormatter("yyyy-MM-dd HH:mm:ss", timeZone: TimeZone(identifier: "UTC")),
        fixedFormatter("HH:mm:ss", timeZone: TimeZone(identifier: "UTC")),
        fixedFormatter("HH:mm", timeZone: TimeZone(identifier: "UTC"))
    ]

    private static let compactDateOut: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US")
        f.timeZone = etTimeZone
        f.dateFormat = "EEE, MMM d"
        return f
    }()

    private static let etTimeOut: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US")
        f.timeZone = etTimeZone
        f.dateFormat = "h:mm a"
        return f
    }()

    /// Format a date string (`YYYY-MM-DD` or ISO 8601) into a compact
    /// `Mon, Sep 14` display, matching RN's `formatCompactDate`.
    static func formatCompactDate(_ raw: String) -> String {
        let parsed: Date? = isoFractional.date(from: raw)
            ?? isoPlain.date(from: raw)
            ?? dateParsers.lazy.compactMap { $0.date(from: raw) }.first
        guard let date = parsed else { return raw }
        return compactDateOut.string(from: date)
    }

    /// Convert a game-time string (UTC ISO or `HH:mm`) to an EST display
    /// string. Mirrors RN's `convertTimeToEST`.
    static func convertTimeToEST(_ raw: String) -> String {
        if raw.isEmpty { return "" }
        let parsed: Date? = isoFractional.date(from: raw)
            ?? isoPlain.date(from: raw)
            ?? timeParsers.lazy.compactMap { $0.date(from: raw) }.first
        guard let date = parsed else { return raw }
        return formatET(date)
    }

    private static func formatET(_ date: Date) -> String {
        etTimeOut.string(from: date) + " ET"
    }

    /// Pick a confidence color for a 50–100% probability slice.
    static func confidenceColor(_ percent: Int) -> Color {
        switch percent {
        case 80...: return Color(red: 0.13, green: 0.77, blue: 0.37) // strong green
        case 70..<80: return Color(red: 0.52, green: 0.80, blue: 0.09) // light green
        case 60..<70: return Color(red: 0.92, green: 0.70, blue: 0.03) // yellow
        default: return Color(red: 0.98, green: 0.45, blue: 0.09) // orange
        }
    }
}

/// Lightweight team-color pair fallback. Replaced per-sport in later
/// batches by the real RN color tables (`getNFLTeamColors`, etc.).
struct TeamColorPair: Hashable, Sendable {
    let primary: Color
    let secondary: Color

    static let neutralNFL = TeamColorPair(primary: Color(hex: 0x002244), secondary: Color(hex: 0xC8102E))
    static let neutralCFB = TeamColorPair(primary: Color(hex: 0x862633), secondary: Color(hex: 0xFFCD00))
    static let neutralNBA = TeamColorPair(primary: Color(hex: 0xC9082A), secondary: Color(hex: 0x17408B))
    /// NCAAB shares CFB's college-sport palette by default. Per-school
    /// gradients ship with #008 alongside CFB / NCAA color tables.
    static let neutralNCAAB = TeamColorPair(primary: Color(hex: 0x862633), secondary: Color(hex: 0xFFCD00))
}

/// Tiny initials helper used as a TeamAvatar fallback when logo URLs aren't
/// wired up yet. Strips parens, takes first letters of up to 2 words.
enum TeamInitials {
    static func from(_ teamName: String) -> String {
        let cleaned = teamName.replacingOccurrences(of: "()", with: "").trimmingCharacters(in: .whitespaces)
        let words = cleaned.split(separator: " ")
        if words.count >= 2 {
            return String(words.prefix(2).map { $0.first ?? "?" })
        }
        return String(cleaned.prefix(3)).uppercased()
    }

    /// Split team into city + nickname pair if the name has multiple words.
    /// Mirrors RN's `getTeamParts(name)`.
    static func parts(of teamName: String) -> (city: String, name: String) {
        let words = teamName.split(separator: " ")
        if words.count < 2 { return (teamName, "") }
        if words.count == 2 {
            return (String(words[0]), String(words[1]))
        }
        // Heuristic for 3-word names like "Los Angeles Lakers": last word is nickname.
        let nickname = words.last.map(String.init) ?? ""
        let city = words.dropLast().joined(separator: " ")
        return (city, nickname)
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/gamecards/GameCardFormatting.kt
package com.wagerproof.app.features.gamecards

import androidx.compose.ui.graphics.Color
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.Locale
import kotlin.math.abs
import kotlin.math.roundToInt

/**
 * Shared card formatting + team-color helpers — port of iOS
 * `GameCards/Components/GameCardFormatting.swift` (pure logic, no UI).
 *
 * All numeric/date formatting used by the sport card adapters funnels through
 * here so every league's row reads identically.
 */
object GameCardFormatting {

    private val ET: ZoneId = ZoneId.of("America/New_York")

    /** "+150" / "-180" / "—". */
    fun formatMoneyline(ml: Int?): String {
        if (ml == null) return "—"
        return if (ml > 0) "+$ml" else ml.toString()
    }

    /** Signed spread: whole when integral, else 1dp. */
    fun formatSpread(spread: Double?): String {
        if (spread == null) return "—"
        val rounded = if (spread == spread.toLong().toDouble()) {
            spread.toLong().toString()
        } else {
            String.format(Locale.US, "%.1f", spread)
        }
        return if (spread > 0) "+$rounded" else rounded
    }

    /** O/U display rounded to nearest half-point. */
    fun roundToNearestHalf(value: Double?): Double? {
        if (value == null) return null
        return (value * 2.0).roundToInt() / 2.0
    }

    /** "EEE, MMM d" in ET (e.g. "Sun, Nov 24"). */
    fun formatCompactDate(raw: String?): String {
        val instant = parseInstant(raw) ?: return raw ?: ""
        return DateTimeFormatter.ofPattern("EEE, MMM d", Locale.US)
            .withZone(ET).format(instant)
    }

    /** "h:mm a ET" in ET (e.g. "1:00 PM ET"). */
    fun convertTimeToEST(raw: String?): String {
        val instant = parseInstant(raw) ?: return raw ?: ""
        val t = DateTimeFormatter.ofPattern("h:mm a", Locale.US).withZone(ET).format(instant)
        return "$t ET"
    }

    /** 4-tier confidence palette shared with `confidenceColor` on iOS. */
    fun confidenceColor(percent: Double): Color = when {
        percent >= 80 -> Color(0.13f, 0.77f, 0.37f)
        percent >= 70 -> Color(0.52f, 0.80f, 0.09f)
        percent >= 60 -> Color(0.92f, 0.70f, 0.03f)
        else -> Color(0.98f, 0.45f, 0.09f)
    }

    /**
     * Tolerant timestamp parse: ISO8601 (±fractional), "yyyy-MM-dd HH:mm:ss",
     * "yyyy-MM-dd". Returns null when unparseable.
     */
    fun parseInstant(raw: String?): Instant? {
        val s = raw?.trim().orEmpty()
        if (s.isEmpty()) return null
        runCatching { OffsetDateTime.parse(s).toInstant() }.getOrNull()?.let { return it }
        runCatching { Instant.parse(s) }.getOrNull()?.let { return it }
        runCatching {
            LocalDateTime.parse(s, DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"))
                .toInstant(ZoneOffset.UTC)
        }.getOrNull()?.let { return it }
        for (pattern in listOf("HH:mm:ss", "HH:mm")) {
            runCatching {
                LocalTime.parse(s, DateTimeFormatter.ofPattern(pattern))
                    .atDate(LocalDate.of(2001, 1, 1))
                    .toInstant(ZoneOffset.UTC)
            }.getOrNull()?.let { return it }
        }
        runCatching {
            LocalDate.parse(s, DateTimeFormatter.ofPattern("yyyy-MM-dd"))
                .atStartOfDay(ET).toInstant()
        }.getOrNull()?.let { return it }
        return null
    }
}

/** Swift `TeamColorPair` — primary/secondary brand colors for a team. */
data class TeamColorPair(val primary: Color, val secondary: Color)

/**
 * City/nickname split helpers — port of iOS `TeamInitials`.
 */
object TeamInitials {

    /** First letters of the first two words, or the first three letters. */
    fun from(name: String): String {
        val cleaned = name.replace("()", "").trim()
        val words = cleaned.split(Regex("\\s+")).filter { it.isNotEmpty() }
        return if (words.size >= 2) {
            words.take(2).mapNotNull { it.firstOrNull() }.joinToString("").uppercase(Locale.US)
        } else {
            cleaned.take(3).uppercase(Locale.US)
        }
    }

    /**
     * (city, nickname) split. 3-word names ("New York Knicks") put the first
     * two words as city, last as nickname; otherwise first/rest.
     */
    fun parts(name: String): Pair<String, String> {
        val words = name.trim().split(" ").filter { it.isNotEmpty() }
        return when {
            words.isEmpty() -> "" to ""
            words.size == 1 -> words[0] to words[0]
            words.size == 2 -> words[0] to words[1]
            else -> words.dropLast(1).joinToString(" ") to words.last()
        }
    }
}
